Fix invalid value error for paymentMethodMessaging in product details page

mgascam opened the pull request:
Fixes #6819 (reopened)

This issue has been reopened. We received reports of errors in the plugin in the product details page after the previous fix was released. This time, errors seem to happen due to merchants customizing this page by removing the quantity input.

#### Changes proposed in this Pull Request

- Improved the method of fetching the product quantity and handling cases where the input element might be missing.
- Introduced safeguards against potential undefined values for productVariations and related fields.
- Established 'USD' as the fallback currency in case it's not defined in productVariations.
- Added more strict checks for potentially missing DOM elements to prevent script errors.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
To replicate the issue we need to remove the quantity input from the product details page. This can be achieved by several means. We got reports of merchants using a page builder. I was able to replicate the issue locally, using this hook `woocommerce_is_sold_individually`. In the functions.php file of the Storefront theme add the following:

```
function custom_woocommerce_remove_quantity_input() {
    return true;
}
add_filter( 'woocommerce_is_sold_individually', 'custom_woocommerce_remove_quantity_input' );
```
* Enable Affirm or Afterpay in your local store.
* Navigate to the details page of a product with variations.
* Check the Affirm and Afterpay messaging is displayed when the page loads.
* Check there are no errors in the browser console related to `paymentMethodMessaging`.
* Check you are able to select variations with no errors happening.
* Check you are able to add the product to the cart without issues.
* Navigate to the details page of a product **without** variations.
* Check the same steps as before.
* Enable the quantity selector back (you can return false in the `custom_woocommerce_remove_quantity_input`)
* Repeat the same steps to check you are able to add the product to the cart.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

mgascam committed:
Fix invalid value error for paymentMethodMessaging

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7153` or branch name `fix/6819-payment-method-messaging-invalid-value` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/6819-payment-method-messaging-invalid-value) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 3df98365f83b00c3a6b434878933adf4f042de7b
- Build time: 2023-09-28 07:55:31 UTC

_Note: the build is updated when a new commit is pushed to this PR._

mgascam committed:
Add changelog

mgascam committed:
Merge branch 'develop' into fix/6819-payment-method-messaging-invalid-value

kalessil commented on client/product-details/index.js:
```diff
@@ -37,13 +49,23 @@ jQuery( function ( $ ) {
-			const quantity = $( '.quantity input[type=number]' ).val();
-			const variationPrice =
-				productVariations[ variation.variation_id ].amount;
+			const quantityInput = $( '.quantity input[type=number]' );
```
I believe we shouldn't engage in updating the element if the input is missing or variation price and currency are not accessible.

mgascam replied:
Done here https://github.com/Automattic/woocommerce-payments/pull/7153/commits/23da3b2d9a1c97f84b4de2c74bcd82774e53d9ef thanks

kalessil commented on client/product-details/index.js:
```diff
@@ -7,27 +7,39 @@
 	$( '.quantity input[type=number]' ).on( 'change', function ( event ) {
-		const newQuantity = event.target.value;
-		const price = productVariations[ productId ].amount;
+		const newQuantity = parseInt( event.target.value, 10 ) || 1;
```
Also here: if we couldn't parse new quantity or variation price and currency are not accessible, we shouldn't update the element.

mgascam replied:
Done here https://github.com/Automattic/woocommerce-payments/commit/23da3b2d9a1c97f84b4de2c74bcd82774e53d9ef thanks

kalessil commented on client/product-details/index.js:
```diff
@@ -7,27 +7,39 @@
 
 	const resetBnplPaymentMessage = () => {
-		const quantity = $( '.quantity input[type=number]' ).val();
+		const quantityInput = $( '.quantity input[type=number]' );
```
Also here: if we couldn't parse new quantity or variation price and currency are not accessible, we shouldn't update the element.

mgascam replied:
Done here https://github.com/Automattic/woocommerce-payments/commit/23da3b2d9a1c97f84b4de2c74bcd82774e53d9ef thanks

kalessil commented on client/product-details/index.js:
```diff
@@ -7,27 +7,39 @@
 import { initializeBnplSiteMessaging } from './bnpl-site-messaging';
 
 jQuery( function ( $ ) {
+	// Check for required global variables
```
Good one, can we expand a bit what's contained in this variable for readers?

mgascam replied:
Done here https://github.com/Automattic/woocommerce-payments/pull/7153/commits/3ad3c2aaf61dcddb961168d7eadfeae460dd7ac9

kalessil reviewed (changes requested):
Like the direction, but we shouldn't be updating element if we can not reliable calculate price. See my comments.

mgascam committed:
Merge branch 'develop' into fix/6819-payment-method-messaging-invalid-value

mgascam committed:
PR feedback: prevent updating if missing quantity or currency

mgascam committed:
PR feedback: improve comment

mgascam committed:
Fall back to quantity 1 when selector is not present

mgascam committed:
Merge branch 'develop' into fix/6819-payment-method-messaging-invalid-value

mgascam committed:
Revert "Fall back to quantity 1 when selector is not present"

This reverts commit 9bc87d5542ab9a385b4322fac43df40b8defde52.

mgascam committed:
Merge branch 'develop' into fix/6819-payment-method-messaging-invalid-value

kalessil commented on client/product-details/index.js:
```diff
@@ -7,51 +7,95 @@
-	let { productId } = window.wcpayStripeSiteMessaging;
+
+	// Utility function to safely parse integers and handle NaN
+	const safeParseInt = ( value ) => {
```
Naming suggession: `parseIntOrReturnZero`

kalessil replied:
How much efforts is it to convert file into TS?

KarlisJ replied:
I'm in favor of existing `safeParseInt` naming

mgascam replied:
I finally renamed to `parseIntOrReturnZero` for being more explicit of what the function does. https://github.com/Automattic/woocommerce-payments/pull/7153/commits/15d2d7d9d496dd3f5f60de7c92f249cf56e6c348 thanks!

> How much efforts is it to convert file into TS?

We are dealing with jQuery so we would need to install the type definitions and also modify the `tsconfig.json`. Nothing major, but I'd rather work on it in a different pull request.

kalessil commented on client/product-details/index.js:
```diff
@@ -7,51 +7,95 @@
-				quantity,
-			currency: productVariations.base_product.currency,
-		} );
+		if ( ! quantityInput.length || ! productVariations.base_product )
```
We are missing `{}` here.

mgascam replied:
Done here https://github.com/Automattic/woocommerce-payments/pull/7153/commits/15d2d7d9d496dd3f5f60de7c92f249cf56e6c348 thanks!

kalessil commented on client/product-details/index.js:
```diff
@@ -7,51 +7,95 @@
+		);
+
+		const amount = baseProductAmount * quantity;
+		const currency = productVariations.base_product.currency;
```
Would it make sense using `productVariations.base_product?.currency` here?

mgascam replied:
Yep, done here https://github.com/Automattic/woocommerce-payments/pull/7153/commits/768537975c36b22b97068a2d490fde8b270b1b12 thanks!

kalessil commented on client/product-details/index.js:
```diff
@@ -7,51 +7,95 @@
+				if (
+					! quantityInput.length ||
+					! productVariations[ variation.variation_id ]
+				)
```
We are missing `{}` here.

mgascam replied:
Done https://github.com/Automattic/woocommerce-payments/pull/7153/commits/15d2d7d9d496dd3f5f60de7c92f249cf56e6c348 thanks!

kalessil commented on client/product-details/index.js:
```diff
@@ -7,51 +7,95 @@
+	// Update BNPL message based on the quantity change
+	quantityInput.on( 'change', ( event ) => {
+		const newQuantity = safeParseInt( event.target.value );
+		const price = safeParseInt( productVariations[ productId ].amount );
```
Would it make sense using `productVariations[ productId ]?.amount` here?

mgascam replied:
Done here https://github.com/Automattic/woocommerce-payments/pull/7153/commits/15d2d7d9d496dd3f5f60de7c92f249cf56e6c348

kalessil commented on client/product-details/index.js:
```diff
@@ -7,51 +7,95 @@
+
+				const quantity = safeParseInt( quantityInput.val() );
+				const variationPrice = safeParseInt(
+					productVariations[ variation.variation_id ].amount
```
Would it make sense using `productVariations[ productId ]?.amount` here?

mgascam replied:
Done https://github.com/Automattic/woocommerce-payments/pull/7153/commits/15d2d7d9d496dd3f5f60de7c92f249cf56e6c348

kalessil reviewed (approved):
Pre-approving, as handed further reviews to @KarlisJ.

KarlisJ reviewed (commented):
Overall, when looking strictly at the changes and comparing old vs new code it's a nice improvement, I like the changes.

But my first though seeing the `resetBnplPaymentMessage` and `quantityInput.on( 'change',` was "why can't we trigger the `change` event from `resetBnplPaymentMessage`?" . I believe that's the "jQuery approch" for less code duplication. 

On the surface level it seems that both blocks operate with different values, but digging deeper it seems to me that it's not true. 

In `resetBnplPaymentMessage` we access `productVariations.base_product.amount` but in the change handler we access `productVariations[ productId ].amount`. Looking [at the source ](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments-payment-method-messaging-element.php#L83) `productId` that we define at the top will always be `base_product` so essentially both blocks perform the same action. 

Or have I misunderstood something?

mgascam committed:
Address PR feedback

mgascam committed:
Merge branch 'fix/6819-payment-method-messaging-invalid-value' of github.com:Automattic/woocommerce-payments into fix/6819-payment-method-messaging-invalid-value

mgascam committed:
Merge branch 'develop' into fix/6819-payment-method-messaging-invalid-value

mgascam committed:
Merge branch 'fix/6819-payment-method-messaging-invalid-value' of github.com:Automattic/woocommerce-payments into fix/6819-payment-method-messaging-invalid-value

mgascam committed:
Address PR feedback

mgascam committed:
Refactor to avoid code repetition

mgascam committed:
Handle quantity change when variation is selected

mgascam committed:
Fix wrong informed amount with variations

mgascam committed:
Fix messaging not updated when quantity selector does  not exist for variations

mgascam committed:
Merge branch 'develop' into fix/6819-payment-method-messaging-invalid-value

mgascam commented:
> Overall, when looking strictly at the changes and comparing old vs new code it's a nice improvement, I like the changes.
> 
> But my first though seeing the `resetBnplPaymentMessage` and `quantityInput.on( 'change',` was "why can't we trigger the `change` event from `resetBnplPaymentMessage`?" . I believe that's the "jQuery approch" for less code duplication.
> 
> On the surface level it seems that both blocks operate with different values, but digging deeper it seems to me that it's not true.
> 
> In `resetBnplPaymentMessage` we access `productVariations.base_product.amount` but in the change handler we access `productVariations[ productId ].amount`. Looking [at the source ](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments-payment-method-messaging-element.php#L83) `productId` that we define at the top will always be `base_product` so essentially both blocks perform the same action.
> 
> Or have I misunderstood something?

Thanks @KarlisJ I looked into your comment about variations and refactored the code. Also fixed some buggy scenarios where the messaging element was not being updated. Please take a look and let me know if you like it better.

> But my first though seeing the `resetBnplPaymentMessage` and `quantityInput.on( 'change',` was "why can't we trigger the `change` event from `resetBnplPaymentMessage`?" . I believe that's the "jQuery approch" for less code duplication.

Apologies but I'm not following you here. `resetBnplPaymentMessage` is a function we use as event handler, not sure what you mean when you suggest to trigger the change event from there?

mgascam committed:
Amend comments

mgascam committed:
Rename variable

mgascam committed:
Merge branch 'develop' into fix/6819-payment-method-messaging-invalid-value

mgascam committed:
Merge branch 'develop' into fix/6819-payment-method-messaging-invalid-value

mgascam committed:
Merge branch 'develop' into fix/6819-payment-method-messaging-invalid-value

KarlisJ commented:
@mgascam 
> Apologies but I'm not following you here. resetBnplPaymentMessage is a function we use as event handler, not sure what you mean when you suggest to trigger the change event from there?

Before your last commits it seemed that both `resetBnplPaymentMessage` and event handler perform the same action. So it seemed to me that we are duplicating logic I suggested to simply call `resetBnplPaymentMessage` in the event handler to avoid duplication.

Now I see that they do not perform same action thus can't be reused is such way.

KarlisJ commented on client/product-details/index.js:
```diff
@@ -7,51 +7,124 @@
+				}
+
+				updateBnplPaymentMessage(
+					productVariations[ variation.variation_id ]?.amount,
```
Given the check above I think it's safe to remove the `?`

mgascam replied:
Yep, removed it https://github.com/Automattic/woocommerce-payments/pull/7153/commits/1cac1e43bbc5ef2ab4354126e123fd2f0a47be19 thanks!

KarlisJ reviewed (approved):
LGTM, approving with one minor comment.

mgascam committed:
PR feedback

mgascam committed:
Merge branch 'develop' into fix/6819-payment-method-messaging-invalid-value

mgascam committed:
Merge branch 'develop' into fix/6819-payment-method-messaging-invalid-value
